client/hotrod-client/src/main/java/org/infinispan/client/hotrod/near/LinkedMapNearCache.java

```diff
@@ -0,0 +1,92 @@
+package org.infinispan.client.hotrod.near;
+
+import org.infinispan.client.hotrod.VersionedValue;
+import org.infinispan.client.hotrod.configuration.NearCacheConfiguration;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * {@link java.util.LinkedHashMap} based near cache implementation.
+ * Concurrent access is controlled by a reentrant RW lock.
+ */
+public class LinkedMapNearCache<K, V> implements NearCache<K, V> {
+
+   private final LinkedHashMap<K, VersionedValue<V>> cache;
+
+   private final ReadWriteLock rwlock = new ReentrantReadWriteLock();
+
+   protected LinkedMapNearCache(LinkedHashMap<K, VersionedValue<V>> cache) {
+      this.cache = cache;
+   }
+
+   @Override
+   public void put(K key, VersionedValue<V> value) {
+      Lock lock = rwlock.writeLock();
+      try {
+         lock.lock();
+         cache.put(key, value);
```

Comment:
According to the LinkedHashMap javadoc, reinsertion does not cause the insertion order to be changed, therefore the eviction handled by removeEldestEntry() will possibly evict a recently refreshed entry. Not really a problem, but a possible optimization for the future. Did you think about initializing the LinkedHashMap to use accessOrder ? Would it degrade performance ?